core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -118,4 +119,62 @@ class ReplicaManagerTest {
 
     TestUtils.verifyNonDaemonThreadsStatus(this.getClass.getName)
   }
+
+  @Test
+  def testClearPurgatoryOnBecomingFollower() {
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect)
+    props.put("log.dir", TestUtils.tempRelativeDir("data").getAbsolutePath)
+    val config = KafkaConfig.fromProps(props)
+    val zkClient = EasyMock.createMock(classOf[ZkClient])
+    val zkUtils = ZkUtils(zkClient, false)
+    val mockLogMgr = TestUtils.createLogManager(config.logDirs.map(new File(_)).toArray)
+    val time = new MockTime()
+    val jTime = new JMockTime
+    val metrics = new Metrics
+    val rm = new ReplicaManager(config, metrics, time, jTime, zkUtils, new MockScheduler(time), mockLogMgr,
+      new AtomicBoolean(false))
+
+    try {
+      var callbackFired = false
+      def callback(responseStatus: Map[TopicPartition, PartitionResponse]) = {
+        assertEquals("Should give NotLeaderForPartitionException", Errors.NOT_LEADER_FOR_PARTITION.code, responseStatus.values.head.errorCode)
+        callbackFired = true
+      }
+
+      val aliveBrokers = Seq(new Broker(0, "host0", 0), new Broker(1, "host1", 1))
+      val metadataCache = EasyMock.createMock(classOf[MetadataCache])
+      EasyMock.expect(metadataCache.getAliveBrokers).andReturn(aliveBrokers).anyTimes()
+      EasyMock.replay(metadataCache)
+
+      val partition = rm.getOrCreatePartition(topic, 0)
+      partition.getOrCreateReplica(0)
+      // Make this replica the leader.
+      val leaderAndIsrRequest1 = new LeaderAndIsrRequest(0, 0,
+        collection.immutable.Map(new TopicPartition(topic, 0) -> new PartitionState(0, 0, 0, Seq(0, 1).asJava.asInstanceOf[java.util.List[Integer]],
+          0, Set(0, 1).asJava.asInstanceOf[java.util.Set[Integer]])).asJava,
```

Comment:
`instanceOf` here can be avoided in the same way as I described a few lines below.